### ISSUE TITLE
PRD: VGGT 3D Encoder + L2/L3 Curriculum Baselines

### ISSUE THREAD

**Issue by LSailer** [LSailer/Master-Thesis-3D#66]
## Problem Statement

The L1 baseline (67% SR, 0.49 SPL with CNN on 64×64 RGB) establishes that R2-Dreamer learns in a single-house setting. Two open questions remain: (1) does the world model degrade on harder curriculum levels (L2: multi-goal, L3: multi-scene)? and (2) does replacing the CNN encoder with 3D geometric features from InfiniteVGGT improve world model performance? Additionally, L1 shows SPL declining over training — a re-run with detailed SPL tracking will diagnose whether the agent learns to find goals but takes increasingly inefficient paths.

## Research Questions

- **RQ1:** Does R2-Dreamer's world model degrade on L2 (6 goals, 1 scene) and L3 (1 goal, 10 scenes) compared to L1?
- **RQ2:** Does a 3D point-cloud representation (VGGT world_points + camera_pose) produce a better world model than 2D RGB features for object navigation?
- **RQ3:** Why does SPL decline during L1 training despite rising SR? Does the agent learn to find goals via longer, less efficient paths?

## Proposed Solution

**Track A — L1/L2/L3 CNN Baselines:** Re-run L1 with detailed SPL tracking (per-episode path length vs shortest path, rolling SPL curve). Run L2 and L3 with same hyperparameters (2.4M steps). Compare SR, SPL, and world model losses across levels.

**Track B — L1 VGGT Encoder:** Replace the CNN encoder with InfiniteVGGT's 3D output. During acting, Habitat renders 518×518 RGB → InfiniteVGGT (PyTorch, streaming mode with KV-cache) produces per-frame `world_points (37×37×3)` + `camera_pose (9)` → flatten + concatenate (4,116 dim) → single linear projection → RSSM embedding. VGGT features are stored in replay; the world model trains on stored features. The KV-cache resets at episode boundaries.

Key design decisions:
- **VGGT-only** — no RGB passed to the agent (clean ablation of 3D vs 2D)
- **Patch-level resolution** (37×37) — skip DPT upsampling, use point head at ViT patch grid
- **Streaming InfiniteVGGT** — one frame per step, KV-cache persists within episode, resets on `done`
- **PyTorch→NumPy→JAX boundary** — VGGT stays in PyTorch, features converted to NumPy before entering JAX agent
- **Single linear projection** — flatten(world_points) + camera_pose → Linear → embed_dim (1024)
- **Barlow Twins** — compares projected RSSM feature against VGGT embedding (unchanged mechanism, now aligns with 3D geometry)

### Module: `VGGTFeatureExtractor`
- **Responsibility**: Wraps InfiniteVGGT for streaming inference during acting. Manages KV-cache lifecycle (reset on episode boundary). Extracts world_points at patch resolution + camera_pose.
- **Interface**:
  ```python
  class VGGTFeatureExtractor:
      def __init__(self, device: str = "cuda") -> None:
          """Load frozen InfiniteVGGT model."""

      def reset(self) -> None:
          """Clear KV-cache (call at episode boundaries)."""

      def extract(self, rgb: np.ndarray) -> dict[str, np.ndarray]:
          """Single-frame streaming inference.
          Args:
              rgb: (3, 518, 518) uint8
          Returns:
              {"world_points": (37, 37, 3) float32,
               "camera_pose": (9,) float32}
          """
  ```
- **Dependencies**: `external/InfiniteVGGT`, PyTorch, CUDA
- **Compute**: GPU (shared with R2Dreamer on same H100)

### Module: `VGGTEncoder` (JAX/Flax)
- **Responsibility**: Replaces `R2Encoder`. Flattens world_points + camera_pose and projects to embed_dim via single linear layer.
- **Interface**:
  ```python
  class VGGTEncoder(nn.Module):
      embed_dim: int = 1024

      def __call__(self, obs: jnp.ndarray) -> jnp.ndarray:
          """Flatten and project VGGT features.
          Args:
              obs: (B, 4116) float32 — flattened world_points (4107) + camera_pose (9)
          Returns:
              embed: (B, embed_dim) float32
          """
  ```
- **Dependencies**: Flax/JAX
- **Compute**: GPU (JAX, negligible vs RSSM)

### Module: `HabitatObjectNavEnv` (modified)
- **Responsibility**: Add 518×518 render resolution option. Return raw RGB for VGGT processing.
- **Interface change**: New config option `render_resolution: int = 64` (set to 518 for VGGT runs).
- **Dependencies**: Habitat-sim

### Module: `ReplayBuffer` (modified)
- **Responsibility**: Store VGGT features instead of RGB. Change obs storage from `(cap, 3, H, W) uint8` to `(cap, 4116) float32`.
- **Interface change**: `add()` and `sample()` accept/return flat feature vectors instead of images. No `/255.0` normalization.
- **Dependencies**: NumPy, JAX

### Module: Training script `run_jax_habitat_vggt.py`
- **Responsibility**: New training script wiring VGGTFeatureExtractor into the acting loop. During acting: env.step → extract VGGT features → store in replay → agent.act on features. During training: unchanged (sample from replay, train world model).
- **Interface**: Same CLI args as `run_jax_habitat.py` + `--encoder vggt`
- **Dependencies**: All above modules
- **WandB**: Log `episode_reset` event markers at episode boundaries for KV-cache reset visibility.

### Module: Enhanced SPL logging (L1 re-run)
- **Responsibility**: Add per-episode `path_length`, `shortest_path`, `path_ratio` (actual/shortest) to WandB + CSV. Rolling SPL curve with windowed breakdown.
- **Interface change**: Additional metrics in training loop logging.
- **Dependencies**: Habitat episode info

### Module: SLURM scripts for L1-rerun/L2/L3
- **Responsibility**: Copy L1 SLURM script, change `--curriculum_path` to L2/L3 JSONs. L1 re-run uses same config with enhanced logging.
- **Dependencies**: Existing `run_jax_habitat.py`

### Track A (L1-rerun/L2/L3 Baselines)
```
Habitat (64×64 RGB) → R2Encoder CNN → embed (1024) → RSSM → policy
                                                          ↓
                                              replay buffer (RGB uint8)
```

### Track B (L1 VGGT)
```
Habitat (518×518 RGB) → InfiniteVGGT (PyTorch, streaming)
                              ↓
                   world_points (37×37×3) + camera_pose (9)
                              ↓ .cpu().numpy()
                   flatten → (4116,) float32
                              ↓
                   replay buffer (features float32)
                              ↓
                   VGGTEncoder (Linear) → embed (1024) → RSSM → policy
```

## SLURM Requirements

| Job Type | Partition | GPUs | Time | Notes |
|----------|-----------|------|------|-------|
| L1 Re-run | gpu_h100 | 1 | 24h | Enhanced SPL logging |
| L2 Baseline | gpu_h100 | 1 | 24h | Same as L1 |
| L3 Baseline | gpu_h100 | 1 | 24h | Same as L1 |
| L1 VGGT | gpu_h100 | 1 | 48h | VGGT adds ~100ms/step during acting |

## Evaluation Criteria

- **Primary metrics**: Success Rate (SR), SPL
- **SPL analysis**: per-episode path_length vs shortest_path, rolling path_ratio, SPL vs SR correlation over training
- **World model diagnostics**: dyn loss (train + val), reward loss, KL divergence
- **Baselines**:
  - L1 CNN: 67% SR, 0.49 SPL (existing)
  - L1 CNN re-run: new (this PRD, with SPL analysis)
  - L2/L3 CNN: new (this PRD)
  - L1 VGGT: new (this PRD)

## Acceptance Criteria

- [ ] L1 re-run completes 2.4M steps with enhanced SPL logging (path_length, shortest_path, path_ratio per episode)
- [ ] L2 baseline completes 2.4M steps, metrics logged to WandB + CSV
- [ ] L3 baseline completes 2.4M steps, metrics logged to WandB + CSV
- [ ] L1 VGGT training runs end-to-end for 2.4M steps without crashes
- [ ] VGGT features (world_points 37×37×3 + camera_pose 9) stored in replay buffer
- [ ] InfiniteVGGT KV-cache resets at episode boundaries, reset events visible in WandB
- [ ] Barlow Twins loss operates on VGGT embeddings (not RGB CNN)
- [ ] Reporter compares L1/L2/L3 CNN baselines and L1 CNN vs L1 VGGT
- [ ] SPL analysis: reporter shows path_ratio trend, SPL vs SR correlation
- [ ] All experiments reproducible via SLURM scripts with documented seeds

## Related Issues

- #63 — Fallback: CNN encoder on VGGT world_points if MLP underperforms
- #64 — Offline replay analysis with OVEL-min dataset
- #65 — Custom semantic head on VGGT aggregated_tokens
